src/pytorch_ood/detector/gram.py

```
@@ -0,0 +1,249 @@
+                        temp = (temp.sign() * torch.abs(temp) ** (1 / p)).reshape(
+                            temp.shape[0], -1
+                        )
+
+                        temp = temp.tolist()
+                        for feature, label in zip(temp, label_list):
+                            if isinstance(feature_class[label][layer_idx][pole_idx], type(None)):
+                                feature_class[label][layer_idx][pole_idx] = feature
+                            else:
+                                feature_class[label][layer_idx][pole_idx].extend(feature)
+                # print update steps
+                if n % 100 == 0:
+                    print(f"Step {n}/{len(data_loader)}")
+
+            for label in range(self.num_classes):
+                for layer_idx in range(self.num_layer):
+                    for poles_idx in range(num_poles):
+                        feature = torch.tensor(
+                            np.array(feature_class[label][layer_idx][poles_idx])
+                        )
+                        current_min = feature.min(dim=0, keepdim=True)[0]
+                        current_max = feature.max(dim=0, keepdim=True)[0]
+
+                        if mins[label][layer_idx][poles_idx] is None:
+                            mins[label][layer_idx][poles_idx] = current_min
+                            maxs[label][layer_idx][poles_idx] = current_max
+                        else:
+                            mins[label][layer_idx][poles_idx] = torch.min(
+                                current_min, mins[label][layer_idx][poles_idx]
+                            )
+                            maxs[label][layer_idx][poles_idx] = torch.max(
+                                current_min, maxs[label][layer_idx][poles_idx]
+                            )
+            self.feature_min = mins
+            self.feature_max = maxs
+            return self
+
+    def fit_features(self: Self, *args, **kwargs) -> Self:
```

this should throw an exception